src/runner/mod.rs

```diff
@@ -0,0 +1,224 @@
+        let l1_provider = Provider::try_from(&self.config.l1_rpc_url)?;
+        let l2_provider = Provider::try_from(&self.config.l2_rpc_url)?;
+        let l2_trusted_provider = Provider::try_from(
+            &self
+                .config
+                .l2_trusted_rpc_url
+                .clone()
+                .expect("a trusted L2 rpc url is required for checkpoint sync"),
+        )?;
+
+        let checkpoint_hash = match self.checkpoint_hash {
+            Some(ref checkpoint) => checkpoint
+                .parse()
+                .expect("invalid checkpoint block hash provided"),
+            None => {
+                tracing::info!("using the latest epoch boundary as checkpoint block");
+
+                let mut block_number = BlockNumber::Latest;
+                loop {
+                    self.check_shutdown()?;
+
+                    let l2_block = l2_trusted_provider
+                        .get_block_with_txs(block_number)
+                        .await?
+                        .unwrap_or_default();
+
+                    let set_l1_block_values_tx = l2_block
+                        .transactions
+                        .iter()
+                        .find(|tx| tx.to.unwrap() == self.config.chain.l1_block)
+                        .expect("could not find setL1BlockValues tx in the fetched block");
+
+                    let sequence_number = &set_l1_block_values_tx
+                        .clone()
+                        .input
+                        .into_iter()
+                        .skip(132)
+                        .take(32)
+                        .collect::<Vec<u8>>();
+
+                    if H256::from_slice(sequence_number) != H256::zero() {
+                        block_number = BlockNumber::Number(
+                            l2_block.number.expect("fetched block has no number") - 1,
+                        );
+                        continue;
+                    }
+
+                    break l2_block.hash.expect("fetched block has no hash");
+                }
+            }
+        };
+
+        tracing::info!("using checkpoint block {}", checkpoint_hash);
+
+        let engine_api = EngineApi::new(&self.config.l2_engine_url, &self.config.jwt_secret);
+        while !engine_api.is_available().await {
+            self.check_shutdown()?;
+            sleep(Duration::from_secs(3)).await;
+        }
+
+        // if the checkpoint block is already synced, start from the finalized head
+        if l2_provider.get_block(checkpoint_hash).await?.is_some() {
+            tracing::warn!("finalized head is above the checkpoint block");
+            self.start_driver().await?;
+            return Ok(());
+        }
+
+        // this is a temporary fix to allow execution layer peering to work
+        // TODO: use a list of whitelisted bootnodes instead
+        tracing::info!("adding trusted peer to the execution layer");
+        l2_provider.add_peer(TRUSTED_PEER_ENODE.to_string()).await?;
```

Comment:
Should I leave this in? It's probably not needed now as you've added op-geth nodes with the right config

Reply:
I think we should leave this for now. We can remove once there are more nodes that have peering enabled.

Reply:
Cool!